### PR TITLE
Extraction runs: the durable run model

### PR DESCRIPTION
**Pull request by jimador**
PR 6 of the extraction-integrity train (refs #67 — the model half), stacked on #94. The durable extraction-run record, as pure JVM value types. `ExtractionRun` is keyed by (`ContextId`, `ExtractionRunRef`) and carries profile, ordered source revisions, pass, parent/superseded refs plus a denormalized root-run ref (OpenLineage's ParentRunFacet pattern — a whole-lineage lookup becomes one indexed read), experiment/cohort refs, prompt/schema/metamodel fingerprints, runtime identity, status, timing, counts, and bounded failures. Requested and observed model facts are separate types with disjoint fields, so a guessed provider-effective value has nowhere to live. Invocation identity is the ordinal in a call plan fixed before any call fires; attempts number retries; completion order never mints. Opaque pseudonymous refs carry actor/request/session identity with serialized-row privacy tests.

**Changed in this review round:**

- **Failures speak a closed vocabulary.** `ExtractionFailure` has no `String` parameter, property or field anywhere in its reachable shape and no factory taking a `Throwable`: `code` (11 values), `stage`, a validated `providerStatus`, one typed `measure` with its unit, `at`, and the invocation it names. `detail`, `fromThrowable` and the sanitize/cause-chain helpers are removed, so prompt content, PII, credentials and protected text have nothing shaped to hold them and cannot reach durable storage through this type.
- **Run recording aligns with the source-revision contract.** The revision bound this type imposed is gone — a run accepts whatever `SourceRevisionRef` accepts, so extraction and revision queries can no longer succeed where run recording then fails.
- **The lineage root is derived.** The constructor is private and `childOf` derives the root from the parent's own lineage, so a root that contradicts its parent has no public path in; the KDoc records the reflection residual honestly.
- `metamodelFingerprint` keeps its field and its KDoc names its one writer: the extraction coordinator, resolving the host's `DeclaredSchemaSource` — per the decision that per-proposition schema attribution answers through run lineage.

**Breaking changes: none.** New types only. The failure-vocabulary rework reshapes types introduced earlier in this same unreleased train; no released surface moves.

**Opt-in and status:** **EXPERIMENTAL** — every type carries `@ApiStatus.Experimental`. Nothing here executes; the store contract, the Drivine store, and lineage stamping follow as #98, #99 and #101.

```mermaid
erDiagram
    ExtractionRun ||--o{ ExtractionInvocationRecord : "invocations, identity fixed before any call"
    ExtractionRun ||--o{ ExtractionFailure : "bounded, typed vocabulary, no free text"
    ExtractionRun ||--|| ExtractionRunLineage : "runRef, root, parent, pass"
    ExtractionRun }o--|| ExtractionContentProfileRef : "profile"
    ExtractionRun }o--o{ SourceRevisionRef : "ordered source revisions"
    ExtractionFailure }o--o| ExtractionInvocationRecord : "may name only an attempt the run records"
```

The denormalized root is what makes a whole-lineage lookup one indexed read.

### CODE REVIEW

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRun.kt [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+            "sourceRevisions must be distinct; a run reads each source revision once"
+        }
+        // SourceRevisionRef predates the run model's cap rule and validates non-blank only, so the
+        // bound is applied where the run stores it. Moving it onto the type is a follow-up.
+        this.sourceRevisions.forEach { revision ->
+            require(revision.sourceKey.length <= ExtractionRunLimits.MAX_SOURCE_KEY_LENGTH) {
+                "sourceKey must be at most ${ExtractionRunLimits.MAX_SOURCE_KEY_LENGTH} characters, " +
+                    "was ${revision.sourceKey.length}"
+            }
+            require(revision.sourceRevision.length <= ExtractionRunLimits.MAX_IDENTIFIER_LENGTH) {
```
Comment (jimador) [inline dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRun.kt]: `SourceRevisionRef` accepts revisions longer than this, so extraction and revision queries can succeed and run recording then fails. Can we keep the bound aligned with the source revision contract until a real store constraint requires one?

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRunLineage.kt [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ * @property rootRunRef The oldest run in this lineage, which is [runRef] itself when there is no
+ *   parent
+ * @property parentRunRef The run this one continues from, or null
+ * @property supersedesRunRef The run this one replaces, or null
+ * @property passIndex Which pass over the material this is, counting from zero
+ */
+@ApiStatus.Experimental
+data class ExtractionRunLineage @JvmOverloads constructor(
+    val runRef: ExtractionRunRef,
+    val rootRunRef: ExtractionRunRef,
```
Comment (jimador) [inline dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRunLineage.kt:71]: Any non-self root passes this constructor, and the store never verifies it against the parent. Can we derive the root from the stored parent or defer this denormalization until a query needs it?

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRun.kt (hunk 1 above)
Comment (jimador) [reply dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRun.kt]: Dropped the cap. `sourceKey` and `sourceRevision` carry no length check here or on `SourceRevisionRef`: a revision is opaque and non-blank per `docs/design/source-revisions.md`, and a run accepts whatever that contract accepts. A bound can return when a real store constraint requires one.

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRunLineage.kt (hunk 2 above)
Comment (jimador) [reply dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRunLineage.kt:71]: The constructor is private and `childOf` derives the root from the parent's own lineage, so a root disagreeing with the parent chain has no public path in. The KDoc records the residual: Kotlin reflection and a Jackson deserializer still reach the private constructor, and nothing serializes this type today.

**Comment by jimador** [comment]
`ExtractionFailure` is not sanitized by construction. Its public constructor and `of` factory accept arbitrary source text, and the Drivine mapper persists that text verbatim. Truncating it and flattening line breaks does not prevent prompt content, PII, credentials, or protected text from entering durable storage.

Make the persisted diagnostic a controlled structured vocabulary, with any detailed material represented by a protected reference. Add canary tests proving raw source text and secrets cannot cross this boundary.

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionFailure.kt [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+ * What kind of thing went wrong, from a fixed list.
+ *
+ * The code is what a query groups by and what an alert fires on, so it comes from a fixed list
+ * that no caller can extend at a call site. Anything that does not fit is [INTERNAL] or
+ * [UNCLASSIFIED], and a code that keeps getting used for the wrong thing is a signal to add one.
+ *
+ * EXPERIMENTAL. The shape may still change while extraction runs (DICE #67) land.
+ */
+@ApiStatus.Experimental
+enum class ExtractionFailureCode {
```
Comment (jimador) [inline dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionFailure.kt:31]: `ExtractionFailure` was not sanitized by construction. Its public constructor and `of` factory accepted arbitrary source text, and the Drivine mapper persisted that text verbatim. Truncating it and flattening line breaks does not prevent prompt content, PII, credentials, or protected text from entering durable storage.

Make the persisted diagnostic a controlled structured vocabulary, with any detailed material represented by a protected reference. Add canary tests proving raw source text and secrets cannot cross this boundary.

Comment (jimador) [reply dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionFailure.kt:31]: Failures speak a closed vocabulary. `ExtractionFailure` has no `String` parameter, property or field anywhere in its reachable shape and no factory taking a `Throwable`: `code` (11 values), `stage`, a validated `providerStatus`, one typed `measure` with its unit, `at`, and the invocation it names. `detail`, `fromThrowable` and the sanitize/cause-chain helpers are removed, so prompt content, PII, credentials and protected text have nothing shaped to hold them and cannot reach durable storage through this type. The Drivine mapper persists the vocabulary (#99, "Persist the failure vocabulary and announce transitions from Drivine").

File: docs/design/extraction-runs.md [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+    EXP --> OP
+```
+
+`ExtractionRunKey` is the tenant-qualified identity: a run id is host-minted and DICE never
+assumes it is unique across tenants, so the `ContextId` travels with it. Two tenants that both
+mint `run-1` have two runs.
+
+## How a run joins to the metamodel
+
+One field carries the join. `ExtractionRunFingerprints.metamodelFingerprint` holds the declared
```
Comment (jimador) [inline docs/design/extraction-runs.md:50]: The join with the metamodel stack was undefined. `ExtractionRun.fingerprints.metamodelFingerprint` was a pass-through string. On the other stack `dice.metamodel.version` was a proposition metadata key nothing writes. Both describe the same fact, which schema declaration a claim was extracted under. Decide once. The run carries the `DeclaredSchema` stamp hash, the coordinator resolves it from the `DeclaredSchemaSource` bean when one is present, and per-proposition schema attribution is answered through `PRODUCED_BY_RUN`. The metadata key then goes, or becomes a denormalised copy the coordinator writes.

Then: the decision was in KDoc only. `ExtractionRunEnvelope.kt` said the coordinator writes `metamodelFingerprint` from the host's `DeclaredSchemaSource`, and this doc never mentioned the field. Put the decision here, in the same words as `metamodel-versioning.md` on the other stack, so the two docs describe one mechanism.

Comment (jimador) [reply docs/design/extraction-runs.md:50]: **One attribution mechanism: the run.** On #83 the metadata key is removed ("Answer schema attribution through run lineage"); the run carries the declared schema's `contentHash` and per-proposition attribution answers through `PRODUCED_BY_RUN`. `metamodelFingerprint` keeps its field, and its KDoc names its one writer, the coordinator resolving the host's `DeclaredSchemaSource`, so nothing else may fill it.

7d8a8dd (on #95's branch) adds "How a run joins to the metamodel" to this doc, in the same words as `metamodel-versioning.md`: the coordinator resolves the host's `DeclaredSchemaSource` and stamps `metamodelFingerprint` on the run record.

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionFailure.kt [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+    /** The model or its service could not be reached, or refused to serve the request. */
+    MODEL_UNAVAILABLE,
```
Comment (igordayen) [inline dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionFailure.kt:34]: MODEL_UNAVAILABLE("description")?

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionFailure.kt [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+ * @property value The number
+ */
+@ApiStatus.Experimental
+data class ExtractionFailureMeasure(
+    val quantity: ExtractionFailureQuantity,
+    val value: Long,
+) {
+
+    init {
+        require(value >= 0) { "$quantity must not be negative, was $value" }
```
Comment (igordayen) [inline dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionFailure.kt:167]: same validation pattern comments,  on PR#94

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionInvocationRecord.kt [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+ *
+ * EXPERIMENTAL. The shape may still change while extraction runs (DICE #67) land.
+ *
+ * @property invocationIndex Ordinal of this call in the run's plan, counting from zero
+ * @property attempt Which try at that call this is, counting from one
+ */
+@ApiStatus.Experimental
+data class ExtractionInvocationId(
+    val invocationIndex: Int,
+    val attempt: Int,
```
Comment (igordayen) [inline dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionInvocationRecord.kt:43]: is there any external config for max attempts

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionInvocationRecord.kt [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+ * EXPERIMENTAL. The shape may still change while extraction runs (DICE #67) land.
+ *
+ * @property inputTokens Tokens the provider counted on the way in
+ * @property outputTokens Tokens the provider counted on the way out
+ * @property totalTokens The provider's own total, when it gave one
+ * @property cachedInputTokens Input tokens the provider served from its cache
+ * @property reasoningTokens Output tokens the provider attributed to reasoning
+ */
+@ApiStatus.Experimental
+data class ExtractionModelUsage @JvmOverloads constructor(
```
Comment (igordayen) [inline dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionInvocationRecord.kt:113]: Maybe extend Usage (from core)?, Seems conceptually similar

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRequestedModelConfig.kt [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+ * @property maxTokens Requested output-token ceiling
+ * @property presencePenalty Requested presence penalty
+ * @property frequencyPenalty Requested frequency penalty
+ * @property thinkingFingerprint Opaque digest of the reasoning or thinking configuration
+ * @property selectionFingerprint Opaque digest of how the host chose this model and these settings
+ * @property timeout How long the caller was prepared to wait
+ */
+@ApiStatus.Experimental
+data class ExtractionRequestedModelConfig @JvmOverloads constructor(
+    val modelRole: String? = null,
```
Comment (igordayen) [inline dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRequestedModelConfig.kt:65]: please consider embedding LlmHyperparameters

File: docs/design/extraction-runs.md [hunk 10]
```diff
@@ -0,0 +1,404 @@
+# Extraction runs: what produced a claim, recorded without holding what it read
+
+An extraction run is the durable record of one execution: which profile, prompt, schema and
```
Comment (igordayen) [inline docs/design/extraction-runs.md:3]: conceptually similar to what-would-be-episode audit

**Review by igordayen — COMMENTED** [review]
@jimador - looks vey good, few comments to consider, thanks

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionFailure.kt (hunk 5 above)
Comment (jimador) [reply dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionFailure.kt:34]: Kept the enum bare on purpose. The constant's name is the wire value: it is what a query groups by, what the REST surface returns, and what an alert matches on, so it has to be stable and language-neutral. The prose lives in KDoc next to each constant, which is where a reader looks it up and where it can change without touching stored data. A description property would be a second copy of that prose that callers would start rendering, and then we could not reword it freely.

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionFailure.kt (hunk 6 above)
Comment (jimador) [reply dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionFailure.kt:167]: Same answer as on #94: `init` runs for every construction path, including secondary constructors and `copy()`, so the invariant holds for every instance. The rule itself is one line per type; there is nothing yet worth pulling into a shared interface.

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionInvocationRecord.kt (hunk 7 above)
Comment (jimador) [reply dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionInvocationRecord.kt:43]: No, and on purpose in this slice. The record is descriptive: `attempt` says which try this was, and the store keeps one row per `(runId, index, attempt)`. Who retries, and how many times, is the caller's policy (the pipeline's retry settings, or the framework's), and the run store neither enforces a ceiling nor knows one. If a ceiling ever needs to be recorded, it belongs on the run header as an observed fact of the plan, not as configuration of the store. Noted as a follow-up for the coordinator slice.

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionInvocationRecord.kt (hunk 8 above)
Comment (jimador) [reply dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionInvocationRecord.kt:113]: Looked at extending it. Core `Usage` is a final data class and it carries `nativeUsage: Any?`, the SDK object this record exists to not store; it also has no cached-input or reasoning counts. So the record stays its own type and gains `ExtractionModelUsage.from(Usage)` in bb6fdfa0, which maps prompt, completion and total tokens and leaves the two extra counts null.

File: dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRequestedModelConfig.kt (hunk 9 above)
Comment (jimador) [reply dice/src/main/kotlin/com/embabel/dice/proposition/extraction/ExtractionRequestedModelConfig.kt:65]: Done in 1c7d081e. `ExtractionRequestedModelConfig` now implements `LlmHyperparameters`; the six fields already matched the interface member for member, so it is the same stored shape read back as the framework's type, plus `from(LlmOptions)` to build one from what a host handed the model. `thinking` stays a fingerprint because `Thinking` is not a stored value.

File: docs/design/extraction-runs.md (hunk 10 above)
Comment (jimador) [reply docs/design/extraction-runs.md:3]: Agreed, and deliberately so. An episode audit is the consumer-side view of the same thing: what ran, with what, and what came out. The run model is the producer-side record DICE can write without knowing anything about the host's episode concept, and `profile` plus the run lineage are the join points a host would use to fold runs into its own audit. I added a short note in the design doc naming that relationship so the next reader does not have to rediscover it.

**Review by igordayen — APPROVED** [review]
@jimador - looks good, subject to rebase. thanks
